File: android-app/app/src/main/java/lvt/crm/data/auth/TokenStore.kt
package lvt.crm.data.auth

import android.content.Context
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey

class TokenStore(context: Context) : CredentialStore {
    private val credentialLock = Any()
    private var revision = 0L
    private val prefs = EncryptedSharedPreferences.create(
        context,
        "lvt_crm_auth",
        MasterKey.Builder(context).setKeyScheme(MasterKey.KeyScheme.AES256_GCM).build(),
        EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
        EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
    )

    override val accessToken: String?
        get() = synchronized(credentialLock) { prefs.getString(KEY_ACCESS, null) }

    override val refreshToken: String?
        get() = synchronized(credentialLock) { prefs.getString(KEY_REFRESH, null) }

    override fun snapshot(): CredentialSnapshot? = synchronized(credentialLock) {
        val access = prefs.getString(KEY_ACCESS, null)?.takeIf { it.isNotBlank() } ?: return@synchronized null
        val refresh = prefs.getString(KEY_REFRESH, null)?.takeIf { it.isNotBlank() } ?: return@synchronized null
        CredentialSnapshot(access, refresh, revision)
    }

    override fun invalidatePendingWrites(): Long = synchronized(credentialLock) {
        ++revision
    }

    override fun saveIfRevision(
        revision: Long,
        accessToken: String,
        refreshToken: String,
    ): Boolean = synchronized(credentialLock) {
        if (this.revision != revision) return@synchronized false
        persist(accessToken, refreshToken)
        true
    }

    override fun replaceIfCurrent(
        expected: CredentialSnapshot,
        accessToken: String,
        refreshToken: String,
    ): Boolean = synchronized(credentialLock) {
        if (currentSnapshot() != expected) return@synchronized false
        persist(accessToken, refreshToken)
        true
    }

    override fun clearIfRevision(revision: Long): Boolean = synchronized(credentialLock) {
        if (this.revision != revision) return@synchronized false
        clearLocked()
        true
    }

    override fun clearIfCurrent(expected: CredentialSnapshot): Boolean = synchronized(credentialLock) {
        if (currentSnapshot() != expected) return@synchronized false
        clearLocked()
        true
    }

    override fun clear() = synchronized(credentialLock) {
        clearLocked()
    }

    private fun currentSnapshot(): CredentialSnapshot? {
        val access = prefs.getString(KEY_ACCESS, null)?.takeIf { it.isNotBlank() } ?: return null
        val refresh = prefs.getString(KEY_REFRESH, null)?.takeIf { it.isNotBlank() } ?: return null
        return CredentialSnapshot(access, refresh, revision)
    }

    private fun persist(accessToken: String, refreshToken: String) {
        val committed = prefs.edit()
            .putString(KEY_ACCESS, accessToken)
            .putString(KEY_REFRESH, refreshToken)
            .commit()
        check(committed) { "TOKEN_PERSIST_FAILED" }
    }

    private fun clearLocked() {
        ++revision
        check(prefs.edit().clear().commit()) { "TOKEN_CLEAR_FAILED" }
    }

    companion object {
        private const val KEY_ACCESS = "access_token"
        private const val KEY_REFRESH = "refresh_token"
    }
}

File: ios-uikit-lvt/LvtCrmUIKit/TokenStore.swift
import Foundation
import Security

final class TokenStore: CredentialStore, @unchecked Sendable {
    private let lock = NSLock()
    private var revision: Int64 = 0
    private let service = "vn.lvt.crm.auth"
    private let accessAccount = "access_token"
    private let refreshAccount = "refresh_token"

    var accessToken: String? {
        lock.lock(); defer { lock.unlock() }
        return read(account: accessAccount)
    }

    var refreshToken: String? {
        lock.lock(); defer { lock.unlock() }
        return read(account: refreshAccount)
    }

    func snapshot() -> CredentialSnapshot? {
        lock.lock(); defer { lock.unlock() }
        return currentSnapshotLocked()
    }

    func invalidatePendingWrites() -> Int64 {
        lock.lock(); defer { lock.unlock() }
        revision += 1
        return revision
    }

    func saveIfRevision(_ revision: Int64, accessToken: String, refreshToken: String) -> Bool {
        lock.lock(); defer { lock.unlock() }
        guard self.revision == revision else { return false }
        persist(accessToken: accessToken, refreshToken: refreshToken)
        return true
    }

    func replaceIfCurrent(_ expected: CredentialSnapshot, accessToken: String, refreshToken: String) -> Bool {
        lock.lock(); defer { lock.unlock() }
        guard currentSnapshotLocked() == expected else { return false }
        persist(accessToken: accessToken, refreshToken: refreshToken)
        return true
    }

    func clearIfRevision(_ revision: Int64) -> Bool {
        lock.lock(); defer { lock.unlock() }
        guard self.revision == revision else { return false }
        clearLocked()
        return true
    }

    func clearIfCurrent(_ expected: CredentialSnapshot) -> Bool {
        lock.lock(); defer { lock.unlock() }
        guard currentSnapshotLocked() == expected else { return false }
        clearLocked()
        return true
    }

    func clear() {
        lock.lock(); defer { lock.unlock() }
        clearLocked()
    }

    private func currentSnapshotLocked() -> CredentialSnapshot? {
        guard let access = read(account: accessAccount)?.nilIfBlank,
              let refresh = read(account: refreshAccount)?.nilIfBlank else { return nil }
        return CredentialSnapshot(accessToken: access, refreshToken: refresh, revision: revision)
    }

    private func persist(accessToken: String, refreshToken: String) {
        write(account: accessAccount, value: accessToken)
        write(account: refreshAccount, value: refreshToken)
    }

    private func clearLocked() {
        revision += 1
        delete(account: accessAccount)
        delete(account: refreshAccount)
    }

    private func read(account: String) -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
        ]
        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)
        guard status == errSecSuccess, let data = item as? Data else { return nil }
        return String(data: data, encoding: .utf8)
    }

    private func write(account: String, value: String) {
        delete(account: account)
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
            kSecValueData as String: Data(value.utf8),
            kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly,
        ]
        let status = SecItemAdd(query as CFDictionary, nil)
        precondition(status == errSecSuccess, "TOKEN_PERSIST_FAILED")
    }

    private func delete(account: String) {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
        ]
        SecItemDelete(query as CFDictionary)
    }
}

private extension String {
    var nilIfBlank: String? {
        let trimmed = trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }
}
